File: solutions/typescript/triangle/1/triangle.ts
export class Triangle {
  private sides: number[]

  constructor(...sides: number[]) {
    this.sides = sides.sort();
  }

  get isEquilateral(): boolean {
    if ( ! this.isValid() ) return false
    return this.sides[0] === this.sides[1] && this.sides[1] == this.sides[2];
  }

  get isIsosceles(): boolean {
    if ( ! this.isValid() ) return false
    return this.sides[0] === this.sides[1] || this.sides[1] == this.sides[2];
  }

  get isScalene(): boolean {
    if ( ! this.isValid() ) return false
    return ! this.isIsosceles && ! this.isEquilateral;
  }

  private isValid(): boolean {
    return (this.sides.length === 3) && (this.sides[0] > 0) && (this.sides[0] + this.sides[1] >= this.sides[2]);
  }
}

File: solutions/elixir/triangle/1/lib/triangle.ex
defmodule Triangle do
  @type kind :: :equilateral | :isosceles | :scalene

  @doc """
  Return the kind of triangle of a triangle with 'a', 'b' and 'c' as lengths.
  """
  @spec kind(number, number, number) :: {:ok, kind} | {:error, String.t()}
  def kind(a, b, c) when a > 0 and b > 0 and c > 0 do
    case Enum.sort([a, b, c]) do
      [x, y, z] when x + y <= z -> {:error, "side lengths violate triangle inequality"}
      [x, x, x] -> {:ok, :equilateral}
      [x, x, _] -> {:ok, :isosceles}
      [_, x, x] -> {:ok, :isosceles}
      [_, _, _] -> {:ok, :scalene}
    end
  end

  def kind(_, _, _), do: {:error, "all side lengths must be positive"}
end
